### PR TITLE
[dtensor] remove torchgen function schema and parse manually

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90241
* __->__ #90106
* #89970
* #89969
* #89968
* #89967

This PR get rids of torchgen FunctionSchema parsing and parse
it manually, it should resolve torchgen package issue and also
provide some perf wins when running DTensor eagerly

### CODE REVIEW

File: torch/distributed/_tensor/dispatch.py
Comment (awgu) [inline torch/distributed/_tensor/dispatch.py:78]: Conceptual: Does this PR regress the computation of `self.is_inplace` and `self._is_out_variant` in some cases because you are using a heuristic instead of `SchemaKind`?

I am curious how you think about the tradeoff of getting rid of the `torchgen` dependency versus possibly having incorrect values for `is_inplace` and `is_out_variant` in some cases (if at all).

Comment (wanchaol) [reply torch/distributed/_tensor/dispatch.py:78]: Good point, from the current operator set registered, i think it should work fine as all of the op signature naming following the same convention for inplace and out variant (`_` and `.out`), so I think this heuristic is relatively safe to use and should not regress existing registered operators. However, it does rely on operators following the same convention of naming, which I think it might get changed. Given that ATen op set is relatively stable at this point, I think we can go with this approach as it's simple and very fast without parsing FunctionSchema, if ever op signature gets changed, the heuristic need to be adapted to accommodate.

File: torch/distributed/_tensor/dispatch.py
Comment (awgu) [inline torch/distributed/_tensor/dispatch.py:294]: General comment: This `return` assumes `len(out_dts) >= 1` or else there will be an index out of bounds error when accessing `out_dts[0]`. I am guessing that this `len(out_dts) >= 1` invariant always holds, but I just wanted to point this out. Maybe you want an assert?

**Review by awgu — COMMENTED** [review]
I am not sure I have the context to approve, so I will defer to someone else.

**Review by awgu — APPROVED** [review]
This looks good to me, so I will stamp to unblock. Feel free to wait for others' review.

**Review by fduwjj — COMMENTED** [review]
Overall this looks good to me and do we have an estimate of potential perf win?

File: torch/distributed/_tensor/dispatch.py
Comment (wanchaol) [reply torch/distributed/_tensor/dispatch.py:294]: I'll land this as is as I don't want to trigger another full CI run, the assert is added in the follow up PR https://github.com/pytorch/pytorch/pull/90241

Comment (wanchaol) [reply torch/distributed/_tensor/dispatch.py:294]: nvm, might still need to change sth, will update.

**Comment by wanchaol** [comment]
> Overall this looks good to me and do we have an estimate of potential perf win?

I don't yet as I didn't profile this, but my suspection is that it could result in around 20-30 perf wins on DTensor execution, feel free to try it out, I'll focus on making other parts be faster if possible.

**Pull request closed without merging**
